### PR TITLE
Small wins in Engine CI job execution

### LINKED ISSUE #7630: Reorder steps in Engine tests job
The current state is, e.g., here - https://github.com/enso-org/enso/actions/runs/5909853023/job/16030854871

Currently, the `./run backend ci-check` does roughly these steps:
- `sbt -Dbench.compileOnly=true all buildEngineDistribution engine-runner/assembly engine-runner/buildNativeImage verifyLicensePackages runtime/Benchmark/compile language-server/Benchmark/compile searcher/Benchmark/compile; runtime/bench`
  - https://github.com/enso-org/enso/actions/runs/5909853023/job/16030854871#step:10:643
  - We don't want to run `runtime/bench` here - i.e., no benchmark dry runs.
- `enso --run test/Benchmarks`.
  - https://github.com/enso-org/enso/actions/runs/5909853023/job/16030854871#step:10:3452
  - This should be the very last step, along with `runtime/bench`.
- `sbt -Dbench.compileOnly=true set Global/parallelExecution:=false; test`
  - https://github.com/enso-org/enso/actions/runs/5909853023/job/16030854871#step:10:3995
- `enso --no-ir-caches --run test/Examples_Tests` ... and other tests

**The optimal order should be:**
- NativeImage build
- Unit tests (`sbt test`)
- Enso tests
- dry-run benchmarks

The reason for that is because if a job fails, it most likely fails in Unit tests, or in NativeImage build step, and not in bench dry-run step.

### PR DESCRIPTION
**Pull request by hubertp**
### Pull Request Description

- (dry-run) benchmarks are run only on Linux - requested by the libs team to reduce CI build time.
- Reorder steps of the Engine Tests job so that dry-run benchmarks are the last step.

### CODE REVIEW

**Comment by hubertp** [comment]
My abilities to code in Rust are close to none so apologies if this feels wrong. It seems to compile though.

File: build/build/src/engine/context.rs [hunk 1]
```diff
@@ -414,13 +376,91 @@ impl RunContext {
...
+            }
+            // We want benchmarks to run only after the other build tasks are done, as they are
+            // really CPU-heavy.
+            let build_command = (!tasks.is_empty()).then_some(Sbt::concurrent_tasks(tasks));
+            let empty_benchmarks = BTreeSet::new();
+            let benchmark_tasks0 = if TARGET_OS == OS::Windows {
+                &empty_benchmarks
+            } else {
+                &self.config.execute_benchmarks
+            };
```
Comment (jdunkerley) [inline build/build/src/engine/context.rs]: Could we flip this so Mac OSX not running too?

```suggestion
            let benchmark_tasks0 = if TARGET_OS == OS::Linux {
                &self.config.execute_benchmarks
            } else {
                &empty_benchmarks
            };
```

**Comment by Akirathan** [comment]
Looking at the last [Engine tests job](https://github.com/enso-org/enso/actions/runs/6249851103/job/16967434733?pr=7846), the reordering was successful - [unit and Enso tests](https://github.com/enso-org/enso/actions/runs/6249851103/job/16967434733?pr=7846#step:10:1089) run before [benchmarks dry run](https://github.com/enso-org/enso/actions/runs/6249851103/job/16967434733?pr=7846#step:10:12945)

File: build/build/src/engine/context.rs (hunk 1 above)
Comment (Akirathan) [reply build/build/src/engine/context.rs]: Sure, done in c53f5e7a656e6aaeee04637be8a4537179c131ec

**Comment by Akirathan** [comment]
On GUI CI / Build Backend (windows), [Clean Before](https://github.com/enso-org/enso/actions/runs/6259306813/job/17000281575?pr=7846) step is failing. I have tried to restart it one more time, and still can see the same error. Let's remove the `CI: Clean build required` label now, as its unnecessary anymore.

**Review by mwu-tow — COMMENTED** [review]
I am wary of disabling some checks just because they are slow. We have a history of letting regressions this way (e.g. mac builds). If this really does help to solve some issue then let's do it, but I'd like to confirm:
1) Is this job actually the slowest one, actually blocking PRs? Aren't IDE jobs slower?
2) Does this change actually provide a significant reduction of time? This is a pretty random and unreliable check but the Engine CI run on this branch took 1 hour 10 minutes, while the last run on the develop tip was 8 minutes faster. Well, it is quite impossible that running fewer things takes more time… but if the change is within the margin of usual variability, than is this really worth it?
3) And in general, I wonder why dry-running benchmarks take much time. Wasn't this dry-run mode introduced just for the purpose of keeping it fast? Is there a way to increase the speed of this check rather than disabling it?

**Comment by JaroslavTulach** [comment]
> We have a history of letting regressions this way (e.g. mac builds).
> If this really does help to solve some issue then let's do it

Disabling of benchmarks dry run on some OSes is fine as it is OS independent. It is enough to run the check on a single OS (e.g. Linux), we don't need to repeat it on other OSes. Linux engine CI build is fastest (faster than Windows and Mac), so it makes sense to keep the test in that one and remove them from the others.

**Comment by jdunkerley** [comment]
@mwu-tow the goal of the dry run is to check that they compile and can be run as we had historically been ignoring them to the point that they were completely out of sink.

They run one iteration rather than the full set and this does take a significant block of time.

Reordering them to the end also means we should have failed before them if any unit test fails which will improve the cycle time for us.

**Comment by mwu-tow** [comment]
I've added a minor change, so we set the config appropriately, rather than selectively ignore sites entry.
If this works and is fine by you, this can land.

**Review by mwu-tow — APPROVED** [review]

**Comment by Akirathan** [comment]
> I've added a minor change, so we set the config appropriately, rather than selectively ignore sites entry. If this works and is fine by you, this can land.

Yes, that is a good idea. Thank you. Merging it right now.

**Pull request merged**